### ISSUE TITLE
Building plugins from multiple private repos

### ISSUE THREAD

**Issue by yroc92** [caddyserver/xcaddy#118]
I have a use-case where I want to build Caddy with plugins from multiple disparate private repos (different accounts / organizations). As an example of why this might be useful, consider the scenario when xcaddy is used in an automated build environment where users can link their private repo via Oauth to create a custom Caddy build.

Retrieving a plugin from a single private repo is trivial (just set `GOPRIVATE` and do a replace on the git remote url host, or use SSH). Multiple private plugins from different organizations would require you to make the the aforementioned changes to the environment just before each `go get` [call](https://github.com/caddyserver/xcaddy/blob/25c412f1ec1ff8efa1987830d7859d32134a98a8/environment.go#L153).

I'm not sure the best way to solve this. My idea is to have `xcaddy` accept a JSON file as a param. It could include an array of "private repo configs" which each consist of the following

- User name
- Password / token
- Repo url

xCaddy would use this info to 

1.  Set `GOPRIVATE` env var (used by `go get` to know when a repo is private),
2. Assemble a temporary `.netrc` before each `go get` call for a private plugin (the `.netrc` is used by the `git` command to authenticate with the git server.

The json config could be nice because would be simple to manage in a CI/CD environment or in your Dockerfile.

Caveat: Right now I'm only thinking about private Github repos, so there might be some minor nuances to supporting / authenticating with other Git servers. 

[Here's an article](https://www.mlo.io/blog/2020/04/14/using-private-go-modules/) about `go get` and private repos for more context.

**Comment by francislavoie** [comment]
Could we do a weird syntax like `github.com/org/repo=!<user>:<pass>@<repo>`? Basically `!` as an early flag for the replacement thing to instead trigger private repo "mode" for the `go get` 🤔

**Comment by yroc92** [comment]
Github at least has [an official syntax the does what you're describing](https://stackoverflow.com/questions/42148841/github-clone-with-oauth-access-token/66156992#66156992)
```
https://<token>@github.com/owner/repo.git
```
I wonder if we should just have xcaddy support that?

**Comment by francislavoie** [comment]
Can you try with `go env -w GOPRIVATE=github.com/org/repo`? Will it work if you just do that? That should override your env to set that for all subsequent Go commands, which should make `xcaddy` use it as well.

**Comment by yroc92** [comment]
It does, but I'd still need to authenticate with my token.

**Comment by francislavoie** [comment]
Right, then I guess that's what the `.netrc` is for.

**Comment by yroc92** [comment]
Right, but imagine you need to authenticate to several private repos because your build requires several private plugins. Then the `.netrc` would need to change before each `go get` call that caddy makes as it iterates through the requested plugins.

**Comment by francislavoie** [comment]
You can have multiple accounts in a netrc file, can't you? I've never used it before but a quick Google suggests you should be able to do that.

**Comment by francislavoie** [comment]
So I assume that did the trick?

**Comment by yroc92** [comment]
I got sidetracked but this recently became relevant again. The problem with the `.netrc` file is that it only allows for 1 credential. Imagine you want to build Caddy with 2 private plugin repos, each from a different account. With `xcaddy`, the only workaround would be to use the `=<replacement>` feature and clone the repos locally with the `oauth` repo syntax (github lets you use a "personal access token" that you can generate in your settings). You can [see an example here](https://github.com/appmasker/plugin-pipeline/blob/main/config/build-caddy.sh).

In addition to the `.netrc` file and using the `oauth` syntax to clone repos and pointing `xcaddy` to the local copy, you can also authenticate via ssh. You can also combine these methods. For example, [this Dockerfile](https://github.com/appmasker/plugin-pipeline/blob/main/config/Dockerfile) is writing a `.netrc` with some provided credentials and then running the example script above with an `oauth` token to clone repos from a different account.

The problem with the `.netrc` and ssh keys is that you can't have multiple keys for a single host (github.com). How do we scale this without writing / replacing stuff in the filesystem? The safest way, in my opinion, is to clone everything locally and use replacements in the `--with` flags, as seen in my example above.

Here's an idea of an interface to solve this:

A `--credentials` flag that accepts JSON / JSON file like so:

```json
[
  {
    "credential": "<personal access token for account A>",
    "repos": ["github.com/private-repo-a/someplugin1", "github.com/private-repo-a/someplugin2"]
  },
  {
    "credential": "<personal access token for account B>",
    "repos": ["github.com/private-repo-b/someplugin3", "github.com/private-repo-b/someplugin4"]
  }
]
```
and behind the scenes it would be cloning and using replacements.

I don't love this, but after researching this issue for many hours, there's not really anything that scales well 🤷🏿 .

**Comment by mholt** [comment]
I could get behind a JSON config file for these more advanced configs...

**Comment by francislavoie** [comment]
Okay cool 👍

**Comment by francislavoie** [comment]
So I guess the probable solution here involves having a JSON config as input as instructions to build.

It would take precedence over any `--with` flags if specified. Maybe `--buildconfig buildconfig.json` (and `--buildconfig -` support for `STDIN` I guess).

This would probably take some heavy refactoring; we'd need a new struct that the JSON unmarshals into, and we'd probably want to change `--with` to append into this build instructions struct, after a bit of string parsing.
